openpype/pipeline/load/plugins.py

```diff
@@ -0,0 +1,130 @@
+    order = 0
+    is_multiple_contexts_compatible = False
+
+    options = []
+
+    log = logging.getLogger("SubsetLoader")
+    log.propagate = True
+
+    def __init__(self, context):
+        self.fname = self.filepath_from_context(context)
+
+    @classmethod
+    def get_representations(cls):
+        return cls.representations
+
+    def filepath_from_context(self, context):
+        return get_representation_path_from_context(context)
+
+    def load(self, context, name=None, namespace=None, options=None):
+        """Load asset via database
+
+        Arguments:
+            context (dict): Full parenthood of representation to load
+            name (str, optional): Use pre-defined name
+            namespace (str, optional): Use pre-defined namespace
+            options (dict, optional): Additional settings dictionary
+
+        """
+        raise NotImplementedError("Loader.load() must be "
+                                  "implemented by subclass")
+
+    def update(self, container, representation):
+        """Update `container` to `representation`
+
+        Arguments:
+            container (avalon-core:container-1.0): Container to update,
+                from `host.ls()`.
+            representation (dict): Update the container to this representation.
+
+        """
+        raise NotImplementedError("Loader.update() must be "
+                                  "implemented by subclass")
+
+    def remove(self, container):
+        """Remove a container
+
+        Arguments:
+            container (avalon-core:container-1.0): Container to remove,
+                from `host.ls()`.
+
+        Returns:
+            bool: Whether the container was deleted
+
+        """
+
+        raise NotImplementedError("Loader.remove() must be "
+                                  "implemented by subclass")
+
+    @classmethod
+    def get_options(cls, contexts):
+        """
+            Returns static (cls) options or could collect from 'contexts'.
+
+            Args:
+                contexts (list): of repre or subset contexts
+            Returns:
+                (list)
+        """
+        return cls.options or []
+
+
+class SubsetLoaderPlugin(LoaderPlugin):
+    """Load subset into host application
+    Arguments:
+        context (dict): avalon-core:context-1.0
+        name (str, optional): Use pre-defined name
+        namespace (str, optional): Use pre-defined namespace
+    """
+
+    def __init__(self, context):
+        pass
+
+
+def discover_loader_plugins():
+    import avalon.api
+
+    return avalon.api.discover(LoaderPlugin)
+
+
+def register_loader_plugin(plugin):
+    import avalon.api
+
+    return avalon.api.register_plugin(LoaderPlugin, plugin)
+
+
+def deregister_loader_plugins_path(path):
+    import avalon.api
+
+    avalon.api.deregister_plugin_path(LoaderPlugin, path)
+
+
+def register_loader_plugins_path(path):
+    import avalon.api
+
+    return avalon.api.register_plugin_path(LoaderPlugin, path)
```

Comment:
Is the goal to strip of `avalon` completely? If so, this doesn't feel like the best strategy. Why still register with Avalon in the e/d

Also, having all these `register_loader_plugin`, etc. separate functions that all do basically the same thing - register a plug-in into a single simple plug-in system. Why do we need all these separate functions everywhere? I know it's more explicit but it introduces so much more code overhead to maintain the logic. OpenPype already is a massive beast - and we're now merging in Avalon **and** at the same time growing Avalon's code into something much bigger too. That feels off.

I feel like that at worst we should merge Avalon code and slim down where we can, no? Unless we're now adding actual functionality that we absolutely require? Or are we just adding convenience function but are we keeping the same minimal system in-place and are we refactoring these utility functions to do `openpype.api.register_plugin_path(LoaderPlugin, path)` in the end?

Reply:
> merging in Avalon and at the same time growing 

Purely merging now. All of these functions exist in avalon repo, hence we are keeping them in the first pass. Once that's done we can start trimming and separating into smaller repos is the ultimate step to allow for leaner codebase and better testability. 

Removing the code from avalon and changing it at the same time feels like a road to hell for us though, so we're just being careful here.

Reply:
> Why do we need all these separate functions everywhere?

Well with this functions you can change it to anything at one place.

> I feel like that at worst we should merge Avalon code and slim down where we can, no? 

Not in single PR. It is a massive amount of changes at one time. It would not be doable and reviewable. It's affecting almost each code in OpenPype and it would also require a refactor of OpenPype because everything is affecting everything.